cpp/inc/bond/core/maybe.h

```diff
@@ -14,143 +18,249 @@ namespace bond
+
+    maybe(maybe&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<typename detail::maybe_common<T>>::value
+                                         && std::is_nothrow_move_constructible<allocator_type>::value)
+        : detail::maybe_common<T>(static_cast<detail::maybe_common<T>&&>(that)),
+          allocator_type(static_cast<allocator_type&&>(that))
+    { }
+
+    /// @brief Move constructor
+    maybe(
+        maybe&& that,
+        const allocator_type& alloc)
+        BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<typename detail::maybe_common<T>>::value
+                         && std::is_nothrow_copy_constructible<allocator_type>::value)
+        : detail::maybe_common<T>(static_cast<detail::maybe_common<T>&&>(that)),
+          allocator_type(alloc)
+    { }
+
+    explicit maybe(const T& value)
+        : detail::maybe_common<T>(value),
+          allocator_type(allocator_type { })
+    { }
+
+    explicit maybe(const allocator_type& alloc)
+        : detail::maybe_common<T>(),
+          allocator_type(alloc)
+    { }
+
+    maybe(const T& value, const allocator_type& alloc)
+        : detail::maybe_common<T>(value, alloc),
+          allocator_type(alloc)
+    { }
+
+    maybe& operator=(const maybe&) = default;
 
     /// @brief Assign a value T
     maybe& operator=(const T& value)
     {
-        _value = value;
-        _nothing = false;
+        // TODO: yes/no on passing allocator?
+        this->_value.emplace(value, static_cast<const allocator_type&>(*this));
         return *this;
     }
 
+    /// @brief If this object contains nothing, set it to a default instance
+    /// of T; otherwise, preserve the existing value. Then, return a
+    /// reference to the value.
+    T& set_value()
+    {
+        if (this->is_nothing())
+        {
+            this->_value.emplace(static_cast<const allocator_type&>(*this));
+        }
 
-    /// @brief Convert to constant reference to value
-    /// @throw NothingException if the object contains nothing
-    operator const T&() const
+        return this->_value.get();
+    }
+
+    /// @brief Swap this object with that object
+    void swap(maybe& that)
     {
-        return value();
+        using std::swap;
+        swap(this->_value, that._value);
+        swap(static_cast<allocator_type&>(*this), static_cast<allocator_type&>(that));
     }
 
-private:
-    T       _value;
-    bool    _nothing;
-};
+    /// @brief Compare for equality
+    bool operator==(const maybe& that) const
+    {
+        return this->_value == that._value
+            &&
+        static_cast<const allocator_type&>(*this)
+        == static_cast<const allocator_type&>(that);
+    }
 
+    bool operator!=(const maybe& that) const
+    {
+        return !(*this == that);
+    }
+};
 
-template<typename T>
-inline void swap(maybe<T>& x, maybe<T>& y)
+template<typename T, typename enabled>
+inline void swap(maybe<T, enabled>& x, maybe<T, enabled>& y)
```

Comment:
I think there is no need to mention the `enabled` here. We are not going to use it with anything else but the default (`void`) one.